sycl/doc/extensions/proposed/sycl_ext_oneapi_device_specific.asciidoc

```diff
@@ -0,0 +1,1217 @@
+way.  (E.g. variables with automatic storage duration or objects created via
+`new` are not allowed.)
+
+* The `device_specific` variable must not itself be an array.  The underlying
+  type _T_ may be an array type, but the `device_specific` variable itself must
+  not be an array.
+
+* The `device_specific` variable must not be shadowed by another identifier _X_
+  which has the same name and is declared in an inline namespace, such that the
+  `device_specific` variable is no longer accessible after the declaration of
+  _X_.
+
+* If the `device_specific` variable is declared in a namespace, none of the
+  enclosing namespace names _N_ may be shadowed by another identifier _X_ which
+  has the same name as _N_ and is declared in an inline namespace, such that
+  _N_ is no longer accessible after the declaration of _X_.
+
+[NOTE]
+====
+The expectation is that some implementations may conceptually insert code at
+the end of a translation unit which references each `device_specific` variable
+that is declared in that translation unit.  The restrictions listed above make
+this possible by ensuring that these variables are accessible at the end of the
+translation unit.
+====
+
+The following example illustrates some of these restrictions:
+
+[source, c++]
+----
+#include <sycl/sycl.hpp>
+using namespace sycl::ext::oneapi::experimental;
+
+device_specific<int> a;           // OK
+static device_specific<int> b;    // OK
+inline device_specific<int> c;    // OK
+
+struct Foo {
+  static device_specific<int> d;  // OK
+};
+device_specific<int> Foo::d;
+
+struct Bar {
+  device_specific<int> e;         // ILLEGAL: non-static member variable not
+};                              // allowed
+
+struct Baz {
+ private:
+  static device_specific<int> f;  // ILLEGAL: not publicly accessible from
+};                              // namespace scope
+device_specific<int> Baz::f;
+
+device_specific<int[4]> g;        // OK
+device_specific<int> h[4];        // ILLEGAL: array of "device_specific" not
+                                // allowed
+
+device_specific<int> same_name;   // OK
+namespace foo {
+  device_specific<int> same_name; // OK
+}
+namespace {
+  device_specific<int> same_name; // OK
+}
+inline namespace other {
+  device_specific<int> same_name; // ILLEGAL: shadows "device_specific" variable
+}                               // with same name in enclosing namespace scope
+inline namespace other2 {
+  namespace foo {               // ILLEGAL: namespace name shadows "::foo"
+  }                             // namespace which contains "device_specific"
+                                // variable.
+}
+----
+
+=== Properties for device-specific variables
+
+The `device_specific` class supports several compile-time-constant properties.
+If specified, these properties are included in the `PropertyListT` template
+parameter as shown in this example:
+
+[source,c++]
+----
+using namespace sycl::ext::oneapi::experimental;
```

Comment:
The `device_specific` spec relies on the `properties` spec, which is in the experimental namespace. Shouldn't we move the `properties` spec, first to the supported namespace before moving the `device_specific` spec?